lib/host/wasi_crypto/symmetric/func.cpp

```diff
@@ -0,0 +1,659 @@
+    return Res.error();
+  } else {
+    *Size = *Res;
+  }
+
+  return __WASI_CRYPTO_ERRNO_SUCCESS;
+}
+
+Expect<uint32_t>
+StateEncryptDetached::body(Runtime::Instance::MemoryInstance *MemInst,
+                           int32_t StateHandle, uint32_t OutPtr,
+                           uint32_t OutLen, uint32_t DataPtr, uint32_t DataLen,
+                           uint32_t /* Out */ TagHandlePtr) {
+  assumingExist(MemInst);
+
+  const __wasi_size_t WasiOutLen = OutLen;
+  auto *const Out = MemInst->getPointer<uint8_t *>(OutPtr, WasiOutLen);
+  assumingExist(Out);
+
+  const __wasi_size_t WasiDataLen = DataLen;
+  auto *const Data = MemInst->getPointer<const uint8_t *>(DataPtr, WasiDataLen);
+  assumingExist(Data);
+
+  auto *const TagHandle =
+      MemInst->getPointer<__wasi_symmetric_tag_t *>(TagHandlePtr);
+  assumingExist(TagHandle);
+
+  if (auto Res = Ctx.symmetricStateEncryptDetached(
+          StateHandle, {Out, WasiOutLen}, {Data, WasiDataLen});
+      unlikely(!Res)) {
+    return Res.error();
+  } else {
+    *TagHandle = *Res;
+  }
+
+  return __WASI_CRYPTO_ERRNO_SUCCESS;
+}
+
+Expect<uint32_t> StateDecrypt::body(Runtime::Instance::MemoryInstance *MemInst,
+                                    int32_t StateHandle, uint32_t OutPtr,
+                                    uint32_t OutLen, uint32_t DataPtr,
+                                    uint32_t DataLen,
+                                    uint32_t /* Out */ SizePtr) {
+  assumingExist(MemInst);
+
+  const __wasi_size_t WasiOutLen = OutLen;
+  auto *const Out = MemInst->getPointer<uint8_t *>(OutPtr, WasiOutLen);
+  assumingExist(Out);
+
+  const __wasi_size_t WasiDataLen = DataLen;
+  auto *const Data = MemInst->getPointer<const uint8_t *>(DataPtr, WasiDataLen);
+  assumingExist(Data);
+
+  auto *const Size = MemInst->getPointer<__wasi_size_t *>(SizePtr);
+  if (unlikely(Size == nullptr)) {
+    return __WASI_CRYPTO_ERRNO_ALGORITHM_FAILURE;
+  }
+
+  if (auto Res = Ctx.symmetricStateDecrypt(StateHandle, {Out, WasiOutLen},
+                                           {Data, WasiDataLen})
+                     .and_then(toWasiSize);
+      unlikely(!Res)) {
+    return Res.error();
+  } else {
+    *Size = *Res;
+  }
+
+  return __WASI_CRYPTO_ERRNO_SUCCESS;
+}
+
+Expect<uint32_t> StateDecryptDetached::body(
+    Runtime::Instance::MemoryInstance *MemInst, int32_t StateHandle,
+    uint32_t OutPtr, uint32_t OutLen, uint32_t DataPtr, uint32_t DataLen,
+    uint32_t RawTagPtr, uint32_t RawTagLen, uint32_t /* Out */ SizePtr) {
+  assumingExist(MemInst);
+
+  const __wasi_size_t WasiOutLen = OutLen;
+  auto *const Out = MemInst->getPointer<uint8_t *>(OutPtr, WasiOutLen);
+  assumingExist(Out);
+
+  const __wasi_size_t WasiDataLen = DataLen;
+  auto *const Data = MemInst->getPointer<const uint8_t *>(DataPtr, WasiDataLen);
+  assumingExist(Data);
+
+  const __wasi_size_t WasiRawTagLen = RawTagLen;
+  auto *RawTag = MemInst->getPointer<uint8_t *>(RawTagPtr, WasiRawTagLen);
```

Comment:
`const uint8_t *`

Reply:
This `RawTag` will be modify later, may use original version.